include/zck.h.in

```diff
@@ -58,125 +58,145 @@ typedef struct zckDL zckDL;
 
 typedef size_t (*zck_wcb)(void *ptr, size_t l, size_t c, void *dl_v);
 
+#ifdef _WIN32
+#define ZCK_WARN_UNUSED
+typedef ptrdiff_t ssize_t;
+#else
+#define ZCK_WARN_UNUSED __attribute__ ((warn_unused_result))
+#endif
+
+#ifdef _WIN32
+    #ifdef ZCHUNK_STATIC_LIB
+        #define PUBLIC
+    #else
+        #ifdef ZCHUNK_EXPORTS
+            #define PUBLIC __declspec(dllexport)
+        #else
+            #define PUBLIC __declspec(dllimport)
+        #endif
+    #endif
+#else
+    #define PUBLIC __attribute__((visibility("default")))
+#endif
 /*******************************************************************
  * Reading a zchunk file
  *******************************************************************/
 /* Initialize zchunk context */
-zckCtx *zck_create(void)
-    __attribute__ ((warn_unused_result));
+PUBLIC zckCtx *zck_create(void)
```

Comment:
For Linux and Mac, we don't need `PUBLIC` defined in the header.  Having it defined in the `.c` files is enough.  For compiling in Windows, is it required in the header?  If not, I'd prefer that we leave it out to improve readability.

Reply:
I think you probably don't want a macro called `-DPUBLIC` in the installed headers anyway, since that name can clash with other software.

Windows is a bit weird about symbol names. I'm not entirely sure how it works, but at least sometimes it seems that you must include annotations in the public header that define whether symbols are exported... and that annotation needs to be *different* depending on whether you're building the library, or linking to it. (So, you effectively need two different copies of the header, or a macro that is used internally by your project to make the header function differently.)


On the other hand, it *may* be enough to include `vs_module_defs: 'zchunk.def'` and list all the symbols in that file. Meson will then build the shared library with that def file in accordance with:

https://docs.microsoft.com/en-us/cpp/build/exporting-from-a-dll-using-def-files?view=msvc-170

Reply:
Thanks for the Meson tip, I can look into that! 
For the PUBLIC it's a common pattern in Windows. On export, one needs to specify `__declspec( dllexport )` and on import one needs `__declspec(dllimport)`. For that reason we need it in the header for sure. 
Additionally, if the declaration doesn't have it the compiler complains about "symbol redefined with different linkage".

Reply:
Is PUBLIC really a common pattern? I thought something non-clashing like ZCHUNK_PUBLIC would be common. ;)